crates/bevy_ecs/src/reflect/entity_commands.rs

```diff
@@ -1,10 +1,44 @@
+//! Define commands and traits for inserting and removing reflected components.
+use std::borrow::Cow;
+use std::marker::PhantomData;
+use std::ops::Deref;
+
+use bevy_reflect::{Reflect, TypeRegistry};
+use parking_lot::RwLockReadGuard;
+
 use crate::prelude::Mut;
 use crate::reflect::AppTypeRegistry;
 use crate::system::{Command, EntityCommands, Resource};
 use crate::{entity::Entity, reflect::ReflectComponent, world::World};
-use bevy_reflect::{Reflect, TypeRegistry};
-use std::borrow::Cow;
-use std::marker::PhantomData;
+
+/// A [`Resource`] holding a [`TypeRegistry`].
+///
+/// This is used with the [`ReflectCommandExt`] methods to identify and extract
+/// a type registry to use to insert/remove [`Reflect`] components.
+///
+/// It is implemented for [`AppTypeRegistry`] and any `Resource` that implements
+/// `AsRef<TypeRegistry>`.
+pub trait ReadTypeRegistry: Resource {
+    /// The type holding the [`TypeRegistry`] to read from.
+    type Target<'t>: Deref<Target = TypeRegistry>
+    where
+        Self: 't;
+
+    /// Read the [`TypeRegistry`] from `Self`.
+    fn type_registry(&self) -> Self::Target<'_>;
+}
+impl<T: Resource + AsRef<TypeRegistry>> ReadTypeRegistry for T {
+    type Target<'t> = &'t TypeRegistry where Self: 't;
+    fn type_registry(&self) -> Self::Target<'_> {
+        self.as_ref()
+    }
+}
+impl ReadTypeRegistry for AppTypeRegistry {
+    type Target<'t> = RwLockReadGuard<'t, TypeRegistry> where Self: 't;
+    fn type_registry(&self) -> Self::Target<'_> {
+        self.read()
+    }
+}
```

Comment:
I feel like this doesn't belong to the `entity_command` module. It's mainly used here, but as a concept it's more general and in my opinion should be defined together with `AppTypeRegistry`.